examples/intro/readme.md

```diff
@@ -0,0 +1,53 @@
+# SimCore Programming Intro
+
+This example provides an introduction to programming model offered by SimCore. First the main programming interfaces are introduced and then the two supported approaches for event-based programming are presented and compared.
+
+## Main interfaces
+
+**Simulation** is the main interface of the framework which allows to configure and execute a simulation model. As demonstrated in [callbacks.rs](src/callbacks.rs), it can be instantiated with a user-defined random seed and then used to create simulation contexts and register event handlers for components of user-defined type `Process`, run the simulation and obtain the simulation time. Besides the `step_until_no_events` method, it provides methods for precise stepping through the simulation. It also provides access to the simulation-wide random number generator which is initialized with the user-defined seed to support deterministic simulations.
+
+**SimulationContext** is the interface for accessing the simulation state and emitting events from components. Each component is associated with a uniquely named context which is created via the `Simulation::create_context` method. The context is typically passed to the component's constructor and is stored inside the component as illustrated in [callbacks.rs](src/callbacks.rs). This example also illustrates the use of the stored context to emit the user-defined events `Request` and `Response`, to obtain the current simulation time, and to generate random numbers using the simulation-wide generator.
+
+SimCore allows a user to keep a reference to a component to call it directly. This is illustrated by `proc1_ref` in [callbacks.rs](src/callbacks.rs)). Moving components completely inside the framework and allowing to interact with them only via events or standard interfaces would harm the usability. It would be more cumbersome to emit a special event to `proc1` instead of calling its method. This also allows to easily inspect component states during the simulation. The same observation applies to the interaction between components - if immediate request/response is assumed, it is both more convenient and efficient to interact via direct calls instead of events. For example, a component modeling CPU can be called directly by other components running on the same simulated machine to request a computation. In response, the CPU component can return the request handle and notify the requester via an event when the computation is completed. Therefore, the framework does not restrict the interaction with and between components to happen only via events. This is in contrast to similar but more strict models such as actor model for message passing.
```

Comment:
```suggestion
SimCore allows a user to keep a reference to a component to call it directly. This is illustrated by `proc1_ref` in [callbacks.rs](src/callbacks.rs). Moving components completely inside the framework and allowing to interact with them only via events or standard interfaces would harm the usability. It would be more cumbersome to emit a special event to `proc1` instead of calling its method. This also allows to easily inspect component states during the simulation. The same observation applies to the interaction between components - if immediate request/response is assumed, it is both more convenient and efficient to interact via direct calls instead of events. For example, a component modeling CPU can be called directly by other components running on the same simulated machine to request a computation. In response, the CPU component can return the request handle and notify the requester via an event when the computation is completed. Therefore, the framework does not restrict the interaction with and between components to happen only via events. This is in contrast to similar but more strict models such as actor model for message passing.
```